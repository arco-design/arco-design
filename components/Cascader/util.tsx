import { isArray } from '../_util/is';
import Store from './base/store';
import { CascaderProps } from './interface';

export const ValueSeparator = '__arco_cascader__';

export const SHOW_PARENT = 'parent';
export const SHOW_CHILD = 'child';

export const PANEL_MODE = {
  cascader: 'cascader',
  select: 'select',
};

export function isEmptyValue(value) {
  return !value || (isArray(value) && value.length === 0);
}

export function getConfig(props: CascaderProps) {
  return {
    showEmptyChildren: props.showEmptyChildren,
    changeOnSelect: props.changeOnSelect,
    lazyload: !!props.loadMore,
    fieldNames: props.fieldNames,
    filterOption: props.filterOption,
    showParent:
      props.mode === 'multiple' && !props.changeOnSelect && props.checkedStrategy === SHOW_PARENT,
  };
}

export function getStore(props, value) {
  const tmp = value ? (Array.isArray(value[0]) ? value : [value]) : [];
  return new Store(props.options || [], tmp, getConfig(props));
}

export const transformValuesToSet = (values: string[][]) => {
  const _values = values || [];
  const valuesSet = _values.reduce((set, next) => {
    // 'next' could be a string.
    set.add([].concat(next).join(ValueSeparator));
    return set;
  }, new Set());

  return valuesSet;
};

export const valueInSet = (set, value: string[]) => {
  const _value = value || [];
  return set.has(_value.join(ValueSeparator));
};

export const removeValueFromSet = (set, value: string[]) => {
  const _value = value || [];
  return set.delete(_value.join(ValueSeparator));
};

export const formatValue = (value, isMultiple, store?): string[][] | undefined => {
  let _value = [];
  if (value === undefined) {
    _value = [];
  } else if (isMultiple) {
    _value = value;
  } else {
    _value = [value];
  }

  if (store && store.config.showParent) {
    const checkedNodes = store.getCheckedNodes();
    const valuesSet = transformValuesToSet(checkedNodes.map((node) => node.pathValue));
    const result = [];
    const temp = {};
    _value.map((v) => {
      v.some((_, index, arr) => {
        const curVal = arr.slice(0, index + 1);
        const pass = valueInSet(valuesSet, curVal);
        if (pass && !temp[curVal.join(ValueSeparator)]) {
          result.push(curVal);
          temp[curVal.join(ValueSeparator)] = 1;
        }
        return pass;
      });
    });

    return result;
  }
  return _value;
};

// change check status to false
const deny2Checked = (option) => {
  const deny = (options) => {
    return !Array.isArray(options)
      ? false
      : options.every((item) => {
          if (item._checked || item.disabled) {
            return true;
          }
          return deny(item.children);
        });
  };
  return option._halfChecked && deny(option?.children);
};

export const getMultipleCheckValue = (propsValue, store: Store<any>, option, _checked) => {
  const checked = _checked && deny2Checked(option) ? false : _checked;

  const beforeValueSet = store.getCheckedNodes().reduce((set, node) => {
    set.add(node.pathValue.join(ValueSeparator));
    return set;
  }, new Set());

  option.setCheckedState(checked);
  const checkedNodes = store.getCheckedNodes();
  const currentValue = checkedNodes.map((node) => node.pathValue);
  const currentValueSet = transformValuesToSet(currentValue);

  const newValueSet = new Set();
  return propsValue
    .filter((v) => {
      // v 不在 beforeValueSet 中，说明 v 不包含对应的option。直接返回true，不应该清除掉。
      if (!valueInSet(beforeValueSet, v) || valueInSet(currentValueSet, v)) {
        newValueSet.add(v.join(ValueSeparator));
        return true;
      }
    })
    .concat(
      currentValue.filter((v) => {
        return !valueInSet(newValueSet, v);
      })
    );
};
