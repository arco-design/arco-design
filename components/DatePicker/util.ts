import dayjs, { Dayjs } from 'dayjs';
import { isArray, isDayjs } from '../_util/is';
import { methods } from '../_util/dayjs';

function getFormat(time) {
  return isDayjs(time) && time.format('HH:mm:ss');
}

export function isTimeArrayChange(prevTime: Dayjs[], nextTime: Dayjs[]) {
  return (
    getFormat(prevTime[0]) !== getFormat(nextTime[0]) ||
    getFormat(prevTime[1]) !== getFormat(nextTime[1])
  );
}

export function getAvailableDayjsLength(value) {
  if (!value) {
    return 0;
  }
  if (isArray(value)) {
    if (isDayjs(value[0]) && isDayjs(value[1])) {
      return 2;
    }
    if (!isDayjs(value[0]) && !isDayjs(value[1])) {
      return 0;
    }
    return 1;
  }
  return 0;
}

// https://github.com/react-component/picker/blob/master/src/utils/dateUtil.ts#L234
export function isDisabledDate(cellDate, disabledDate, mode): boolean {
  if (typeof disabledDate !== 'function') {
    return false;
  }
  // Whether cellDate is disabled in range
  const getDisabledFromRange = (
    currentMode: 'date' | 'month' | 'year',
    start: number,
    end: number
  ) => {
    let current = start;
    while (current <= end) {
      let date: Dayjs;
      switch (currentMode) {
        case 'date': {
          date = methods.set(cellDate, 'date', current);
          if (!disabledDate(date)) {
            return false;
          }
          break;
        }
        case 'month': {
          date = methods.set(cellDate, 'month', current);
          if (!isDisabledDate(date, disabledDate, 'month')) {
            return false;
          }
          break;
        }
        case 'year': {
          date = methods.set(cellDate, 'year', current);
          if (!isDisabledDate(date, disabledDate, 'year')) {
            return false;
          }
          break;
        }
        default:
          break;
      }
      current += 1;
    }
    return true;
  };
  switch (mode) {
    case 'date':
    case 'week': {
      return disabledDate(cellDate);
    }
    case 'month': {
      const startDate = 1;
      const endDate = cellDate.endOf('month').get('date');
      return getDisabledFromRange('date', startDate, endDate);
    }
    case 'quarter': {
      const startMonth = Math.floor(cellDate.get('month') / 3) * 3;
      const endMonth = startMonth + 2;
      return getDisabledFromRange('month', startMonth, endMonth);
    }
    case 'year': {
      return getDisabledFromRange('month', 0, 11);
    }
    default:
      return false;
  }
}

type WeekStartType = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export function getDefaultWeekStart(dayjsLocale: string): WeekStartType {
  return (dayjs.Ls?.[dayjsLocale]?.weekStart as WeekStartType) || 0;
}

export function getLocaleDayjsValue(
  date: Dayjs | undefined,
  dayjsLocale: string
): Dayjs | undefined {
  return date ? date.locale(dayjsLocale) : date;
}
