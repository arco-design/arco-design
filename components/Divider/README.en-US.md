`````
Component / Layout

# Divider

Divide the content area and separate the modules.
`````

%%Content%%

## API

### Divider

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|orientation|The position of description content in Divider|'left' \| 'right' \| 'center' |`center`|
|type|Two types are available: `horizontal` and `vertical`|'horizontal' \| 'vertical' |`horizontal`|
|className|Additional css class|string \| string[] |`-`|
|style|Additional style|CSSProperties |`-`|
