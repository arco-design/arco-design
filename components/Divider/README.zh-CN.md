`````
组件 / 布局

# 分割线 Divider

划分内容区域，对模块做分隔。
`````

%%Content%%

## API

### Divider

|参数名|描述|类型|默认值|
|---|---|---|---|
|orientation|分割线文字的位置|'left' \| 'right' \| 'center' |`center`|
|type|分割线的类型，是水平还是竖直，分别对应 `horizontal` 和 `vertical`|'horizontal' \| 'vertical' |`horizontal`|
|className|节点类名|string \| string[] |`-`|
|style|节点样式|CSSProperties |`-`|
