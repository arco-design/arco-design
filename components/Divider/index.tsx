import React, { useContext, forwardRef } from 'react';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import { DividerProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const defaultProps: DividerProps = {
  type: 'horizontal',
  orientation: 'center',
};

function Divider(baseProps: DividerProps, ref) {
  const { getPrefixCls, componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<DividerProps>(baseProps, defaultProps, componentConfig?.Divider);
  const { children, style, className, type, orientation, ...rest } = props;

  const prefixCls = getPrefixCls('divider');
  const classNames = cs(
    prefixCls,
    `${prefixCls}-${type}`,
    {
      [`${prefixCls}-with-text`]: children,
      [`${prefixCls}-with-text-${orientation}`]: children && orientation,
    },
    className
  );

  return (
    <div role="separator" ref={ref} className={classNames} style={style} {...rest}>
      {children && type === 'horizontal' ? (
        <span className={`${prefixCls}-text ${prefixCls}-text-${orientation}`}>{children}</span>
      ) : null}
    </div>
  );
}

const DividerComponent = forwardRef<unknown, DividerProps>(Divider);

DividerComponent.displayName = 'Divider';

export default DividerComponent;

export { DividerProps };
