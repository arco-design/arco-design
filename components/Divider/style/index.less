@import './token.less';

@divider-prefix-cls: ~'@{prefix}-divider';

.@{divider-prefix-cls} {
  &-horizontal {
    position: relative;
    width: 100%;
    min-width: 100%;
    max-width: 100%;
    margin: @divider-margin-vertical 0;
    border-bottom: @divider-size @divider-line-style @divider-color-bg;
    clear: both;

    &.@{divider-prefix-cls}-with-text {
      margin: @divider-margin-vertical_text 0;
      display: flex;
      align-items: center;
      border-bottom-width: 0;

      &::before,
      &::after {
        content: '';
        height: 0;
        flex: 1;
        border-bottom: @divider-size @divider-line-style;
        border-bottom-color: inherit;
      }
    }

    &.@{divider-prefix-cls}-with-text-left:before {
      flex-basis: @divider-position-text-left;
      flex-grow: 0;
    }

    &.@{divider-prefix-cls}-with-text-right:after {
      flex-basis: @divider-position-text-right;
      flex-grow: 0;
    }
  }

  &-vertical {
    display: inline-block;
    min-width: 1px;
    max-width: 1px;
    height: 1em;
    margin: 0 @divider-margin-horizontal;
    border-left: @divider-size @divider-line-style @divider-color-bg;
    vertical-align: middle;
  }

  // 仅水平分割线才支持文字，只考虑水平情况
  &-text {
    box-sizing: border-box;
    padding: 0 @divider-margin-text;
    font-size: @divider-font-size;
    font-weight: @divider-font-weight;
    line-height: 2;
    color: @divider-color-text;
  }
}
