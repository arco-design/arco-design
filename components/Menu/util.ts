import React, { ReactElement } from 'react';

export type MenuInfo = {
  keyPath: string[];
  prev: string;
  next: string;
  firstChild?: string;
  lastChild?: string;
  disabled?: boolean;
};

export const PROPS_NEED_TO_BE_PASSED_IN_SUBMENU = ['popup', 'triggerProps', 'selectable'];

// Expand MenuGroup to get an array only contains MenuItem and SubMenu
const flatMenuGroup = (children): ReactElement[] => {
  let validMenuItems = [];

  React.Children.forEach(children, (item) => {
    const menuType = item?.type?.menuType;
    if (menuType === 'MenuItem' || menuType === 'SubMenu') {
      validMenuItems.push(item);
    } else if (menuType === 'MenuGroup') {
      validMenuItems = validMenuItems.concat(flatMenuGroup(item.props.children));
    }
  });

  return validMenuItems;
};

export const generateInfoMap = (
  children,
  keyPath: string[] = [],
  result: {
    [key: string]: MenuInfo;
  } = {}
) => {
  const validChildrenList: any[] = flatMenuGroup(children);

  validChildrenList.forEach((item, index) => {
    const key = item.key;
    const menuType = item.type.menuType;
    const _keyPath = [key, ...keyPath];
    const info: MenuInfo = {
      keyPath: [],
      prev: validChildrenList[index - 1]?.key || null,
      next: validChildrenList[index + 1]?.key || null,
    };

    if (index === 0 || index === validChildrenList.length - 1) {
      const parentKey = _keyPath[1];
      const propertyName = index === 0 ? 'firstChild' : 'lastChild';
      if (parentKey) {
        result[parentKey] = {
          ...result[parentKey],
          [propertyName]: key,
        };
      }
    }

    switch (menuType) {
      case 'SubMenu':
        info.keyPath = _keyPath;
        generateInfoMap(item.props.children, _keyPath, result);
        break;

      case 'MenuItem':
        info.keyPath = _keyPath;
        info.disabled = item.props.disabled;
        break;

      default:
        break;
    }

    result[key] = {
      ...result[key],
      ...info,
    };
  });

  return result;
};

export const processChildren = (
  children,
  props
): Array<Exclude<unknown, boolean | null | undefined>> => {
  return React.Children.map(children, (item, index) => {
    if (!item || !item.props) {
      return item;
    }

    const isHTMLElement = typeof item.type === 'string';
    const isMenuSubComponent = item.type && item.type.menuType;

    // 处理 <div> 包裹 MenuItem 之类的情况
    if (!isMenuSubComponent && item.props.children) {
      const _props = isHTMLElement ? {} : props;
      return React.cloneElement(item, {
        ..._props,
        _key: item.key,
        children: processChildren(item.props.children, props),
      });
    }

    return isHTMLElement
      ? item
      : React.cloneElement(item, {
          ...props,
          // Properties of the component itself have higher priority
          ...item.props,
          _key: item.key || `$menu-${index}`,
        });
  });
};

export function isChildrenSelected(children, keys: string[]) {
  let find = false;
  function loop(_children) {
    if (!_children || find) {
      return;
    }
    React.Children.forEach(_children, (c) => {
      if (c && c.props && c.type && !find) {
        const menuType = c.type.menuType;
        const selectable = c.props.selectable;
        if (menuType === 'MenuItem') {
          find = keys.indexOf(c.key) !== -1;
        } else if (menuType === 'SubMenu' && selectable) {
          find = keys.indexOf(c.key) !== -1;
        }
        if (!find && c.props.children) {
          loop(c.props.children);
        }
      }
    });
  }
  loop(children);
  return find;
}
