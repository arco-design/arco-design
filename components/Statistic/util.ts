import { padStart } from '../_util/pad';

const units: [string, number][] = [
  ['Y', 1000 * 60 * 60 * 24 * 365], // years
  ['M', 1000 * 60 * 60 * 24 * 30], // months
  ['D', 1000 * 60 * 60 * 24], // days
  ['H', 1000 * 60 * 60], // hours
  ['m', 1000 * 60], // minutes
  ['s', 1000], // seconds
  ['S', 1], // million seconds
];

export function getDateString(millisecond: number, format: string) {
  let leftMillisecond: number = millisecond;
  return units.reduce((current, [name, unit]) => {
    if (current.indexOf(name) !== -1) {
      const value = Math.floor(leftMillisecond / unit);
      leftMillisecond -= value * unit;
      return current.replace(new RegExp(`${name}+`, 'g'), (match: string) => {
        const len = match.length;
        return padStart(value.toString(), len, '0');
      });
    }
    return current;
  }, format);
}
