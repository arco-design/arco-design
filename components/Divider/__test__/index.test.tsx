import React from 'react';
import { render } from '../../../tests/util';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Divider from '..';

mountTest(Divider);
componentConfigTest(Divider, 'Divider');

describe('Divider', () => {
  it('render correctly', () => {
    const divider = render(<Divider />);

    expect(divider.find('.arco-divider').length).toBe(1);
  });

  it('with text', () => {
    const divider = render(<Divider>Hello</Divider>);

    expect(divider.find('.arco-divider-text').length).toBe(1);
    expect(divider.find('.arco-divider-text')[0].textContent).toBe('Hello');
  });

  it('vertical', () => {
    const divider = render(<Divider type="vertical">Hello</Divider>);

    expect(divider.find('.arco-divider-vertical').length).toBe(1);
    expect(divider.find('.arco-divider-text').length).toBe(0);
  });
});
