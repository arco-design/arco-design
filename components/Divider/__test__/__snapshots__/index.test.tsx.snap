// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Divider default 1`] = `
<div
  class="arco-divider arco-divider-horizontal"
  role="separator"
/>
`;

exports[`ConfigProvider componentConfig.Divider set className globally 1`] = `
<div
  class="arco-divider arco-divider-horizontal global-default-class"
  role="separator"
/>
`;

exports[`ConfigProvider componentConfig.Divider set className globally and component set className itself. 1`] = `
<div
  class="arco-divider arco-divider-horizontal component-class"
  role="separator"
/>
`;

exports[`ConfigProvider componentConfig.Divider set data-* property 1`] = `
<div
  class="arco-divider arco-divider-horizontal"
  data-test-name="Divider"
  role="separator"
/>
`;
